src/Microsoft.AspNet.Mvc.Core/ModelBinding/DictionaryModelBinder.cs

```diff
@@ -14,14 +17,86 @@ namespace Microsoft.AspNet.Mvc.ModelBinding
+            {
+                // No match for the prefix at all.
+                return result;
+            }
+
+            var model = result.Model as Dictionary<TKey, TValue>;
+            Debug.Assert(model != null);
+            if (model.Count != 0)
+            {
+                // ICollection<KeyValuePair<TKey, TValue>> approach was successful.
+                return result;
+            }
+
+            var enumerableValueProvider = bindingContext.ValueProvider as IEnumerableValueProvider;
+            if (enumerableValueProvider == null)
+            {
+                // No IEnumerableValueProvider available for the fallback approach.
+                return result;
+            }
+
+            // Attempt to bind dictionary from a set of prefix[key]=value entries. Get the short and long keys first.
+            var keys = await enumerableValueProvider.GetKeysFromPrefixAsync(bindingContext.ModelName);
+            if (!keys.Any())
+            {
+                // No entries with the expected keys.
+                return result;
+            }
+
+            // Update the existing successful but empty ModelBindingResult.
+            var metadataProvider = bindingContext.OperationBindingContext.MetadataProvider;
+            var valueMetadata = metadataProvider.GetMetadataForType(typeof(TValue));
+            var valueBindingContext = ModelBindingContext.GetChildModelBindingContext(
+                bindingContext,
+                bindingContext.ModelName,
+                valueMetadata);
+
+            var modelBinder = bindingContext.OperationBindingContext.ModelBinder;
+            var validationNode = result.ValidationNode;
+
+            foreach (var key in keys)
+            {
+                var dictionaryKey = ConvertFromString(key.Key);
+                valueBindingContext.ModelName = key.Value;
+
+                var valueResult = await modelBinder.BindModelAsync(valueBindingContext);
+
+                // Always add an entry to the dictionary but validate only if binding was successful.
+                model[dictionaryKey] = ModelBindingHelper.CastOrDefault<TValue>(valueResult?.Model);
+                if (valueResult != null && valueResult.IsModelSet)
+                {
+                    validationNode.ChildNodes.Add(valueResult.ValidationNode);
+                }
+            }
+
+            return result;
+        }
+
+        /// <inheritdoc />
         protected override object GetModel(IEnumerable<KeyValuePair<TKey, TValue>> newCollection)
         {
             return newCollection?.ToDictionary(kvp => kvp.Key, kvp => kvp.Value);
         }
 
+        /// <inheritdoc />
         protected override object CreateEmptyCollection()
         {
             return new Dictionary<TKey, TValue>();
         }
+
+        private static TKey ConvertFromString(string keyString)
+        {
+            // Use InvariantCulture to convert string since ExpressionHelper.GetExpressionText() used that culture.
```

Comment:
Not sure this is right - are you saying that when we generate a form-field with this format the key is always `ToString()`ed with invariant?